### ISSUE TITLE
not able to predict

### ISSUE THREAD

**Issue by sahil-dhuri** [ultralytics/yolov5#5394]
i have trained my model in same way as u did for max 50 epochs

but then to its not detecting
![image](https://user-images.githubusercontent.com/58507710/139424278-d032bc3d-8062-4873-b81f-cf625f90dd4c.png)

i have seen in train folder the images are labeled very well
![image](https://user-images.githubusercontent.com/58507710/139424354-d0680738-4359-4abd-b6fd-d97377659630.png)

but it was not able to predict while training itself i think so..
![image](https://user-images.githubusercontent.com/58507710/139424406-6af82464-a54d-4829-966b-4818c65dd046.png)

**Comment by rlrahulkanojia** [comment]
A little more detail regarding the training process would be helpful like the command you used and the summary of the trainng.

**Comment by glenn-jocher** [comment]
@sahil-dhuri 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by sahil-dhuri** [comment]
> 
> 
> @sahil-dhuri 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results.
> 
> Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.
> 
> If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.
> 
> We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.
> ## Dataset
> 
>     * **Images per class.** ≥ 1500 images per class recommended
> 
>     * **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
> 
>     * **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
> 
>     * **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
> 
>     * **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
> 
>     * **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.
> 
> 
> [<img alt="COCO Analysis" width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png">](https://arxiv.org/abs/1405.0312)
> ## Model Selection
> 
> Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models.
> 
> <img alt="YOLOv5 Models" width="700" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png">
> 
>     * **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
> 
> 
> ```shell
> python train.py --data custom.yaml --weights yolov5s.pt
>                                              yolov5m.pt
>                                              yolov5l.pt
>                                              yolov5x.pt
>                                              custom_pretrained.pt
> ```
> 
>     * **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
> 
> 
> ```shell
> python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
>                                                       yolov5m.yaml
>                                                       yolov5l.yaml
>                                                       yolov5x.yaml
> ```
> 
> ## Training Settings
> 
> Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.
> 
>     * **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
> 
>     * **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
> 
>     * **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
> 
>     * **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).
> 
> 
> ## Further Reading
> 
> If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

i tried again for 11 epochs
confusion matrix

![confusion_matrix](https://user-images.githubusercontent.com/58507710/139554029-92d038c7-6a66-4483-918f-70775f550d57.png)

f1 curve

![F1_curve](https://user-images.githubusercontent.com/58507710/139554041-c9468dc0-312d-4e70-99fa-0b160efb944b.png)

labels



![labels](https://user-images.githubusercontent.com/58507710/139554084-b5afa6e1-aa16-4961-b459-c810ae9c662e.jpg)

labels_correlogram

![labels_correlogram](https://user-images.githubusercontent.com/58507710/139554098-57a5d676-620d-422e-8ac3-3fd790088e9a.jpg)

p-curv

![P_curve](https://user-images.githubusercontent.com/58507710/139554105-87480b3e-f380-46ad-a07e-d1de60b17c9c.png)

pr-curv

![PR_curve](https://user-images.githubusercontent.com/58507710/139554109-012767e0-3a53-48d0-b149-24f82b85c6bc.png)

r-curv


![R_curve](https://user-images.githubusercontent.com/58507710/139554119-ae15f553-42b5-44ba-8871-6a47e7f98c5c.png)

results(csv file)

[results.csv](https://github.com/ultralytics/yolov5/files/7447391/results.csv)

results

![results](https://user-images.githubusercontent.com/58507710/139554148-94bc8bc7-bc1d-48de-829b-f64b1ce823e0.png)

trainbatch0

![train_batch0](https://user-images.githubusercontent.com/58507710/139554157-fe1e38f7-f2e6-4100-b5f3-14c023f60d28.jpg)

trainbatch1


![train_batch1](https://user-images.githubusercontent.com/58507710/139554167-f4359997-e9c1-4876-a317-29291e744315.jpg)

trainbatch2

![train_batch2](https://user-images.githubusercontent.com/58507710/139554169-6fe0aa1f-5805-4d3c-a78f-046ec73749e1.jpg)

val_batch0_labels
![val_batch0_labels](https://user-images.githubusercontent.com/58507710/139554184-715833ab-bea4-4b35-81ac-0c5e68d29ff0.jpg)

val_batch0_pred

![val_batch0_pred](https://user-images.githubusercontent.com/58507710/139554194-dd6e55eb-3b5d-4242-8089-3e5d18709bf7.jpg)

THANKS FOR YOUR EARLY REPLY TO MY QUERY I HOPE YOU WILL REPLY SOONER FOR THIS TO BECAUSE I HAVE TO SUBMIT IT AS A PROJECT TILL 2 NOV

**Comment by glenn-jocher** [comment]
@rlrahulkanojia the above guidelines speak for themselves I believe. You should review them and apply them for best results.

**Comment by rlrahulkanojia** [comment]
Hello @glenn-jocher ,
I was just helping out @sahil-dhuri. I think there is a small misunderstanding. 

Thanks for helping

**Comment by sahil-dhuri** [comment]
> 
> 
> @rlrahulkanojia the above guidelines speak for themselves I believe. You should review them and apply them for best results.

tried according to the guide;line but still its not predicting

**Comment by sahil-dhuri** [comment]
> 
> 
> Hello @glenn-jocher , I was just helping out @sahil-dhuri. I think there is a small misunderstanding.
> 
> Thanks for helping

m using ryzen n the second thing is while training the gpu memory is showing 0  any idea for this @glenn-jocher

**Comment by glenn-jocher** [comment]
@sahil-dhuri YOLOv5 will use CUDA resources if available to your environment, defaulting to device 0. If your device 0 is not used by default then your environment is not setup correctly.

Please ensure you meet all dependency requirements if you are attempting to run YOLOv5 locally.  If in doubt, create a new virtual Python 3.8 environment, clone the latest repo (**code changes daily**), and `pip install -r requirements.txt` again. We also highly recommend using one of our verified environments below. 


## Requirements

[**Python>=3.6.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.7**](https://pytorch.org/get-started/locally/). To get started:
```bash
$ git clone https://github.com/ultralytics/yolov5
$ cd yolov5
$ pip install -r requirements.txt
```

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), validation ([val.py](https://github.com/ultralytics/yolov5/blob/master/val.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/export.py)) on MacOS, Windows, and Ubuntu every 24 hours and on every commit.
